### DISCUSSION TITLE
How can I `use` plugins from within a plugin?

### DISCUSSION THREAD

**Discussion by brandoncc** [wbthomason/packer.nvim#479] [General]
Thanks for your work on this awesome plugin manager, which works great!

I moved my config to a plugin so that it is super portable to any nvim install, and I would like to be able to have a single `:PackerSync` install both my plugin and then any plugins that it specifies. Is that possible? I currently have to install, then close nvim, then reopen and sync again. After this, I have all of the plugins I expect.

Here is my current configuration:

https://github.com/brandoncc/nvim-config-as-plugin/blob/main/lua/brandoncc/lv/lv-config.lua (this is the plugin that specifies other plugins)
https://github.com/ChristianChiarulli/LunarVim/blob/master/lua/plugins.lua#L250-L253 (this is where LunarVim brings in my plugin)

This is the configuration I have that sets my LunarVim user_plugins and results in my plugin coming in:

```lua
O.user_plugins = {
  { "brandoncc/nvim-config-as-plugin" }
}
```

I have tried things like this:

```lua
O.user_plugins = {
  { "brandoncc/nvim-config-as-plugin", run = function()
    vim.cmd(':luafile ~/.config/nvim/lv-config.lua')
    vim.cmd(':PackerCompile')
    vim.cmd(':PackerSync')
  end }
}
```

Which successfully ran sync twice, but the sub-plugins were not seen/installed.

Am I trying to do something that is impossible?

**Comment by wbthomason** [comment]
This is an interesting use case! I think you should be able to hack this in reasonably easily, but I will admit that `packer` hasn't been used this way (to my knowledge) before. There's been discussion of adding support for a file specifying plugin requirements to `packer`, but that's a pretty major feature and I haven't gotten around to it yet.

Right now, I think you could do this by:
1. Specify your main plugin in your main `packer` file, using the manual config method (i.e. not the `startup` function)
2. In your main plugin, `require` `packer` and just call `use` with your sub-plugins somewhere in a runtime files (e.g. in `your_plugin/plugin/add_requirements.lua`. Your main plugin will need to be eagerly-loaded (or at least loaded before you use any `packer` commands) for this to get sourced
3. Run `packer` commands as usual and you should see the sub-plugins in the managed set

**Reply by brandoncc** [reply]
Thank you!
